[wasm] rewrite crypto-worker to typescript

pavelsavara opened the pull request:
- rewrite to TS and add type of `InitCryptoMessageData`
- in the crypto-worker we forward the logs to WS based on `config.diagnosticTracing`
- returned some console writes in hope they will prevented the deadlock.


Moved some changes from this PR to https://github.com/dotnet/runtime/pull/73468 and https://github.com/dotnet/runtime/pull/73073

eerhardt commented on src/mono/wasm/runtime/logging.ts:
```diff
@@ -0,0 +1,207 @@
+import Configuration from "consts:configuration";
```
This needs a Copyright header

eerhardt commented on src/mono/wasm/runtime/workers/dotnet-crypto-worker.js:
```diff
@@ -379,16 +380,20 @@ function _stringify_err(err) {
 onmessage = function (p) {
-    var data = p;
+    /**
+     * @type {any}
```
Does this describe `data` or `p`?

Also, is there any value to this comment? Does saying something is `any` have any benefit?

pavelsavara replied:
It would be much better to rewrite the whole thing in proper typeScript, including the type of the message.
I annotated this just to get rid of type warnings in my editor, but in truth, I didn't try to understand all of the code. 
@eerhardt I believe that you guys contributed most of it ?

pavelsavara replied:
I made a guess on the types and converted it to typeScript. I also fixed the `data` vs `event` confusion.

eerhardt commented on src/mono/wasm/runtime/logging.ts:
```diff
@@ -0,0 +1,207 @@
+    }
+}
+
+export function setup_proxy_console(id: string, console: any, origin: string): void {
```
Why is this `any`? Shouldn't it be:

```suggestion
export function setup_proxy_console(id: string, console: Console, origin: string): void {
```

pavelsavara replied:
`console[m] = proxyConsoleMethod` this assignment would need to fully match the type.

pavelsavara replied:
I cast it to any bit later, so that method signature could be better, but the ts compiler is ok to assign to that object.

eerhardt commented on src/mono/wasm/runtime/logging.ts:
```diff
@@ -0,0 +1,207 @@
+                    func([prefix + payload, ...args.slice(1)]);
+                }
+            } catch (err) {
+                originalConsole.error(`proxyConsole failed: ${err}`);
```
This is reference `originalConsole` which doesn't seem to be declared yet.

pavelsavara replied:
We are in the scope of the `proxyConsoleMethod` and the inner function would be called after the `originalConsole` already has a value. I could re-order the code to make it more readable.

radical commented on src/mono/wasm/runtime/workers/dotnet-crypto-worker.js:
```diff
@@ -379,16 +380,20 @@ function _stringify_err(err) {
     if (p.data !== undefined) {
         data = p.data;
     }
+    if (Configuration === "Debug") {
```
This should depend on a setting which flows in from the config, so it can be enabled via the host, xharness etc.

pavelsavara replied:
which setting do you propose ? I did this in order to trim the `Release` size.

pavelsavara replied:
Because this is copy of the code, because it's in separate target js file.

pavelsavara replied:
Also, do we really want to have log forwarding in the production code ? I think it has security implications.

pavelsavara replied:
I did `if (data.config.debug_level)` for now and fixed the startup to call `init_crypto` after the config was loaded already.

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
/azp run runtime-wasm

eerhardt commented on src/mono/wasm/runtime/workers/dotnet-crypto-worker.ts:
```diff
@@ -0,0 +1,386 @@
+                this._throw_if_reset_or_shutdown();
+
+                // Read in request
+                const req: string = this._read_request();
```
Does this need the explicit type? It should be inferred from the method it calls.

```suggestion
                const req = this._read_request();
```

eerhardt commented on src/mono/wasm/runtime/workers/dotnet-crypto-worker.ts:
```diff
@@ -0,0 +1,386 @@
+
+                // Append the latest part of the message.
+                const view = this.msg.slice(0, size_to_read);
+                const part = String.fromCharCode.apply(null, view as any);
```
Why is the `as any` necessary here?

lambdageek commented on src/mono/wasm/runtime/workers/dotnet-crypto-worker.ts:
```diff
@@ -0,0 +1,386 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import { setup_proxy_console } from "../logging";
+import { InitCryptoMessageData } from "../crypto-worker";
```
could be `import type` which might be better to guard that we're not rolling up unwanted code.

lambdageek commented on src/mono/wasm/runtime/workers/dotnet-crypto-worker.ts:
```diff
@@ -0,0 +1,386 @@
+let s_channel;
+
+// Initialize WebWorker
+onmessage = function (event: MessageEvent) {
```
nit: `self.addEventListener("message", (event) => { ... })`

lambdageek reviewed (commented):
thanks for converting the crypto worker to TS!

pavelsavara commented on src/mono/wasm/runtime/logging.ts:
```diff
@@ -0,0 +1,210 @@
+    };
+    consoleWebSocket.onerror = function (event) {
+        originalConsole.error(`[${id}] websocket error: ${event}`, event);
+    };
```
WS has buffering, it's probably better this way.

pavelsavara commented:
I [wrote](https://github.com/dotnet/runtime/issues/72810#issuecomment-1201127819) and then spoke to @javiercn this morning about how to locate the url of the `dotnet-crypto-worker.js` during Blazor startup. How they could influence the location of the file.
We agreed it would be ideal if they could give us the pre-resolved URL as `config.asset`.

I added new resource type `js-module-crypto` and `js-module-threads`, for which Blazor could configure it's asset `resolvedUrl`.
I also changed the detection of blazor startup sequence to not be disabled by presence of `config`, but only by `config.assets` having some `assembly` in it.

@radical I know you guys also spoke about it.
@lambdageek MT will also need asset for it's `dotnet.worker.js` I also included it in this PR too.

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
> Why is this changing from `slice`? I think `slice` reads better.

because slice makes copy, the constructor make a view. Unless I made some mistake.

eerhardt commented:
> because slice makes copy, the constructor make a view. Unless I made some mistake.

Ah, that's a difference between JS and .NET. Looks like what we want is `subarray`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/subarray

lambdageek commented:
> @lambdageek MT will also need asset for it's `dotnet.worker.js` I also included it in this PR too.

I'm not familiar with how the assets stuff works...  Emscripten just does this:

```javascript
 allocateUnusedWorker: function() {
  if (!Module["locateFile"]) {
   PThread.unusedWorkers.push(new Worker(new URL("dotnet.worker.js", import.meta.url)));
   return;
  }
  var pthreadMainJs = locateFile("dotnet.worker.js");
  PThread.unusedWorkers.push(new Worker(pthreadMainJs));
 },
```

but I don't see any changes to `locateFile` in this PR

pavelsavara commented:
> _Update_ I suppose we could replace `PThread.allocateUnusedWorker` with one that goes through the asset mechanism

Right, different PR

pavelsavara commented:
/azp run runtime-wasm

radical commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -628,6 +631,56 @@ function downloadResource(request: ResourceRequest): LoadingResource {
+        switch (behavior) {
+            case "js-module-crypto":
+                asset = {
+                    name: "dotnet-crypto-worker.js",
```
Should we rename these files to follow some pattern like `dotnet.crypto.worker.js`, `dotnet.threads.worker.js`?

pavelsavara replied:
not on this PR, also we I don't know if we could change the name of `dotnet.worker.js` easily until we implement `PThread.init` replacement @lambdageek mentioned.

pavelsavara replied:
tried `-s PTHREAD_WORKER_FILE=dotnet-threads-worker.js` emcc linker options and got
`PTHREAD_WORKER_FILE is an internal setting and cannot be set from command line`

radical commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -628,6 +631,56 @@ function downloadResource(request: ResourceRequest): LoadingResource {
+    return asset;
+}
+
+function resolve_url(asset: AssetEntry, sourcePrefix: string): string {
```
Is this function expected to always return a valid url?

pavelsavara replied:
on node on windows it returns `C:/some/path/file.xxx` on browser `http://some.net/path/file.xxx` on V8 `./file.xxx`

radical commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -842,48 +880,62 @@ let configLoaded = false;
-export async function mono_wasm_load_config(configFilePath: string): Promise<void> {
+export async function mono_wasm_load_config(configFilePath?: string): Promise<void> {
     if (configLoaded) {
+        return afterConfigLoaded.promise;
```
Does this need to do `await afterConfigLoaded.promise` instead?

pavelsavara replied:
that's the same thing. The caller will await it. It doesn't matter which promise.

radical commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -842,48 +880,62 @@ let configLoaded = false;
+        config.assets = config.assets || [];
+        config.runtime_options = config.runtime_options || [];
+        config.globalization_mode = config.globalization_mode || "auto";
+        if (config.enable_debugging)
+            config.debug_level = config.enable_debugging;
+        if (config.debug_level === undefined && BuildConfiguration === "Debug") {
+            config.debug_level = -1;
+        }
```
Isn't this a change in behavior? We are disabling this by default for Debug config?
Also, this seems disconnected from the debug_level stuff in `mono_wasm_load_runtime`.

pavelsavara replied:
```	 * debug_level > 0 enables debugging and sets the debug log level to debug_level
	 * debug_level == 0 disables debugging and enables interpreter optimizations
	 * debug_level < 0 enabled debugging and disables debug logging.
```
so we are enabling it for debug build configuration. Yes that's change in default behavior.

In `mono_wasm_load_runtime` we are just propagating value from `config` if it was not provided as argument.

pavelsavara replied:
Now I did the same also for `diagnosticTracing` on debug build of runtime.

radical commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -842,48 +880,62 @@ let configLoaded = false;
+            config.debug_level = -1;
+        }
+        runtimeHelpers.diagnostic_tracing = !!runtimeHelpers.config.diagnostic_tracing;
+        runtimeHelpers.enable_debugging = runtimeHelpers.config.enable_debugging ? runtimeHelpers.config.enable_debugging : 0;
```
And this also seems disconnected from the debug_level/enable_debugging above?

pavelsavara replied:
I have open different PR on which I'm removing `enable_debugging` completely.

pavelsavara replied:
It gone by now

pavelsavara commented:
The CI failures are unrelated. Is there more feedback ?

pavelsavara commented:
/azp run runtime-wasm

lambdageek commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -261,6 +262,11 @@ async function mono_wasm_pre_init_essential_async(): Promise<void> {
     await init_polyfills_async();
+    await mono_wasm_load_config(Module.configSrc);
+    init_crypto();
+    if (MonoWasmThreads && ENVIRONMENT_IS_PTHREAD) {
+        await mono_wasm_pthread_worker_init();
+    }
```
```suggestion
```

This is gone in `main`

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
/azp run runtime-wasm

radical commented:
`subtlecrypto` tests [failing](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-73073-merge-9dabea0bc60c4d1c81/WasmTestOnBrowser-subtlecrypto-System.Security.Cryptography.Tests/1/console.05ec6af9.log?helixlogtype=result) on Linux:
```
[21:52:18] info: Using random seed for test cases: 874821143
[21:52:18] info: Using random seed for collections: 874821143
[21:52:18] info: Starting:    System.Security.Cryptography.Tests.dll
[22:07:12] fail: Tests timed out. Killing driver service pid 23409
[22:07:12] fail: Application has finished with exit code TIMED_OUT but 0 was expected
XHarness exit code: 71 (GENERAL_FAILURE)
```

pavelsavara commented:
`Expected unchanged file` is https://github.com/dotnet/runtime/issues/73419 
[Log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-73073-merge-0dd4631092454bdcbc/Workloads-.SimpleSourceChangeRebuildTest/1/console.0b68bd9a.log?helixlogtype=result)

radical commented:
@eerhardt have you seen the subtlecrypto tests taking more than 15mins on CI? I haven't seen that before, but Pavel mentioned that the tests are pretty slow. Do we need to bump the xharness timeout for them?

pavelsavara commented:
It could be also some dead-lock. Locally I'm staring at 
![image](https://user-images.githubusercontent.com/271576/183060839-06cd0a8d-c571-4bf3-80ed-227af8981c14.png)

for few minutes now. I undone the change from `slice` to `subarray` we discussed earlier and it didn't help.

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
/azp run runtime-wasm

eerhardt commented:
> have you seen the subtlecrypto tests taking more than 15mins on CI?

No - locally on my codespaces VM they take around 70-90 seconds.

pavelsavara commented:
> > have you seen the subtlecrypto tests taking more than 15mins on CI?
> 
> No - locally on my codespaces VM they take around 70-90 seconds.

On my windows, when it works the it's 2 minutes.
`Total: 2038, Errors: 0, Failed: 0, Skipped: 1, Time: 126.22934s`

But I saw it deadlock multiple times, now I'm trying to reproduce it on main, with just the logging fixed here https://github.com/dotnet/runtime/pull/73468

eerhardt commented:
> now I'm trying to reproduce it on main, with just the logging fixed here 

Yeah, one concern I have with this PR is that you are trying to change too much at once. It is super hard to know what went wrong when making multiple changes in a single PR.

pavelsavara commented:
I was able to reproduce the deadlock on this branch locally on my windows https://github.com/dotnet/runtime/pull/73468
It takes many iterations of
```
dotnet.exe build /p:TargetOS=Browser /p:TargetArchitecture=wasm /p:Configuration=Debug /t:Test src\libraries\System.Security.Cryptography\tests /p:Scenario=WasmTestOnBrowser /p:UseSubtleCryptoForTests=true /p:WasmTestAppArgs="-class System.Security.Cryptography.Encryption.Aes.Tests.AesContractTests" 
```

![image](https://user-images.githubusercontent.com/271576/183129063-979fadb1-ae1a-4976-9925-bf610f4400da.png)

pavelsavara committed:
move changes to other PRs and rebase on main

pavelsavara commented:
> It is super hard to know what went wrong when making multiple changes in a single PR.

Moved some changes from this PR to https://github.com/dotnet/runtime/pull/73468 and https://github.com/dotnet/runtime/pull/73073 and left only rewrite to TS here.

pavelsavara commented:
/azp run runtime-wasm

eerhardt reviewed (approved):
LGTM - thanks for splitting this up!

pavelsavara commented:
@eerhardt are you guys able to have look at the deadlock ? do you need more info?

eerhardt commented:
I haven't been able to look at the deadlock. Does it happen in `main`, or only with your change?

pavelsavara commented:
Probably on main as well, but [the fix to logging](https://github.com/dotnet/runtime/pull/73468) will help you debugging it without XHarness. Otherwise your chrome will crash with OOM.

eerhardt commented:
> Probably on main as well

I haven't seen a test failure report in a CI, and the tests have been running for over a month.

pavelsavara committed:
returning some noisy console logs in hope that they are serializing access and preventing the deadlock. This is RC1 stop-gap.

pavelsavara commented:
Returned some console writes (which I commented out 10 days ago) in hope they will prevented the deadlock. I think they were serializing the access to shared buffer. Perhaps we could find better solution later.

pavelsavara commented:
/azp run runtime-wasm
